### PR TITLE
ci: add changelog generator and commitlint

### PR DESCRIPTION
**Pull request by oldgiova**
Added release-please and git cliff to generate changelogs and Github Releases

Ticket: QA-776

Added optional commit linter: `commitlint`.
* [Sample commit linting failure](https://gitlab.com/Northern.tech/Mender/mender-server/-/jobs/7989782047)
* [Sample commit linting ok](https://gitlab.com/Northern.tech/Mender/mender-server/-/jobs/7989791060)

### CODE REVIEW

**Review by mzedel — APPROVED** [review]

File: .gitlab-ci.yml [hunk 1]
```diff
@@ -230,3 +259,72 @@ coveralls:done:
...
+    # Setting up git
+    - |
+      echo "INFO - setting up git"
+      git config --global user.email "${GITHUB_USER_EMAIL}"
+      git config --global user.name "${GITHUB_USER_NAME}"
+    - npm install -g git-cliff
+    # GITHUB_TOKEN for Github cli authentication
+    - export GITHUB_TOKEN=${GITHUB_CLI_TOKEN}
+    # getting the centralized git cliff config
+    - wget --output-document cliff.toml https://raw.githubusercontent.com/mendersoftware/mendertesting/master/utils/cliff.toml.scoped
```
Comment (alfrunes) [inline .gitlab-ci.yml]: Could we maybe maintain the `cliff.toml` as part of this repository?

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -230,3 +259,72 @@ coveralls:done:
...
+  script:
+    - release-please release-pr
+        --token=${GITHUB_BOT_TOKEN_REPO_FULL}
+        --repo-url=${GITHUB_REPO_URL}
+        --target-branch=${CI_COMMIT_REF_NAME}
+    # git cliff: override the changelog
+    - test $GIT_CLIFF == "false" && echo "INFO - Skipping git-cliff" && exit 0
+    - git clone https://${GITHUB_USER}:${GITHUB_BOT_TOKEN_REPO_FULL}@github.com/${GITHUB_REPO_URL}
```
Comment (alfrunes) [inline .gitlab-ci.yml]: Why are we cloning the repository (again)? Can't we simply add Github as a remote?
```
git remote add upstream https://${GITHUB_USER}:${GITHUB_BOT_TOKEN_REPO_FULL}@github.com/${GITHUB_REPO_URL}
```

**Comment by mzedel** [comment]
I just assume this will be merged after https://github.com/mendersoftware/mendertesting/pull/340

File: release-please-config.json [hunk 3]
```diff
@@ -0,0 +1,16 @@
+{
+  "packages": {
+    ".": {
```
Comment (alfrunes) [inline release-please-config.json:3]: Should there be a package per service?

File: .gitlab-ci.yml (hunk 1 above)
Comment (mzedel) [reply .gitlab-ci.yml]: This file will be used in different repositories and having it in a shared location ensures the changelog will look consistent across our releases.

**Comment by oldgiova** [comment]
> I just assume this will be merged after [mendersoftware/mendertesting#340](https://github.com/mendersoftware/mendertesting/pull/340)

Yes, sorry, I forgot to specify it

File: .gitlab-ci.yml (hunk 1 above)
Comment (oldgiova) [reply .gitlab-ci.yml]: > Could we maybe maintain the `cliff.toml` as part of this repository?

By having a central `cliff.toml`, we'll use a single "standard" for every projects, not just the mender-server, but also clients project. In the end, the central `cliff.toml` is what we want to achieve with: https://northerntech.atlassian.net/browse/QA-763

File: release-please-config.json (hunk 3 above)
Comment (oldgiova) [reply release-please-config.json:3]: No, otherwise we'll end up with different versions

**Comment by oldgiova** [comment]
Also added commitlint; please review again

File: commitlint.config.js [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      'always',
+      {
+        feat: commonScopes,
+        fix: commonScopes,
+        test: commonScopes,
+        perf: [],  // scope is not allowed
+        ci: []     // scope is not allowed
+      }
+    ]
+  },
```
Comment (mzedel) [inline commitlint.config.js:32]: Thank you for this! This is soooo much nicer!

**Review by mzedel — APPROVED** [review]

**Comment by kjaskiewiczz** [comment]
General question - do we need to group changelogs by services (if I understand correctly this is what we are doing/trying to do with current git cliff configuration)?

**Comment by oldgiova** [comment]
> General question - do we need to group changelogs by services (if I understand correctly this is what we are doing/trying to do with current git cliff configuration)?

We have to define it. The very same changelog file will populate also the [documentation website](https://docs.mender.io/release-information/release-notes-changelog/mender-server). What we're achieving here, is also what costumers will see on the doc website.

Right now we're creating a single CHANGELOG.md file with group by `groups`, then group by `scope` (service). For example:

```
# v3.8.0

## Bug fixes
### Deployments
- fix a thing
- fix another thing

### Auditlogs
- fix here
- fix there
- fix everywhere

## Features
### Deployments
- feature here
- feathre there

### Deviceconnect
- feature here
- feature there
```

File: .gitlab-ci.yml [hunk 5]
```diff
@@ -322,3 +349,84 @@ coveralls:done:
...
+        --repo-url=${GITHUB_REPO_URL}
+        --target-branch=${CI_COMMIT_REF_NAME}
+    # git cliff: override the changelog
+    - test $GIT_CLIFF == "false" && echo "INFO - Skipping git-cliff" && exit 0
+    - git remote add github-${CI_JOB_ID} https://${GITHUB_USER_NAME}:${GITHUB_BOT_TOKEN_REPO_FULL}@github.com/${GITHUB_REPO_URL} || true  # Ignore already existing remote
+    - gh repo set-default https://${GITHUB_USER_NAME}:${GITHUB_BOT_TOKEN_REPO_FULL}@github.com/${GITHUB_REPO_URL}
+    - RELEASE_PLEASE_PR=$(gh pr list --author "${GITHUB_USER_NAME}" --head "release-please--branches--${CI_COMMIT_REF_NAME}" --json number | jq -r '.[0].number // empty')
+    - test -z "$RELEASE_PLEASE_PR" && echo "No release-please PR found" && exit 0
+    - gh pr checkout --force $RELEASE_PLEASE_PR
+    - wget --output-document cliff.toml https://raw.githubusercontent.com/mendersoftware/mendertesting/master/utils/cliff.toml
```
Comment (oldgiova) [inline .gitlab-ci.yml:407]: The changelog is standard (non-scoped)

**Pull request merged**